pkg/syscallfilter/syscallfilter_linux.go

```diff
@@ -0,0 +1,206 @@
+		}
+	}
+	switch r.Event {
+	case strace.SyscallEnter:
+		return "E" + sysname
+	case strace.SyscallExit:
+		return "X" + sysname
+	case strace.SignalExit:
+		return fmt.Sprintf("SignalExit")
+	case strace.Exit:
+		return fmt.Sprintf("Exit")
+	case strace.SignalStop:
+		return fmt.Sprintf("SignalStop")
+	case strace.NewChild:
+		return fmt.Sprintf("NewChild")
+	}
+	log.Panicf("Unknown event %#x from record %v", r.Event, r)
+	return ""
+}
+
+func (c *Cmd) handleEvent(t strace.Task, r *strace.TraceRecord, e []*event) error {
+	// All attempts to use defer for printing got ... weird.
+	var ret error
+	n := eventName(r)
+	act := findEvent(e, n)
+	if act == nil {
+		return nil
+	}
+	switch act.Action {
+	case "error":
+		ret = fmt.Errorf("%v", act.Value)
+		if c.Log != nil {
+			fmt.Fprintf(c.Log, "%v act %v error %v\n", n, act.Name, ret)
+		}
+		// And, since we can't really return an error to the process, but only to strace
+		// and, since this is more about killing it than anything else ...
+		// bye bye
+		c.cancel()
+	// Actually, log is the same as error,0 but ... it makes the intent clearer.
+	case "log":
+		if c.Log != nil {
+			fmt.Fprintf(c.Log, "%v act %v error %v\n", n, act.Name, ret)
+		}
+	}
+	return ret
+}
+
+// Run implements cmd.Run, filtering strace events
+// as created by AddActions. The slice can be empty, in which case the command
+// runs as normal.
+func (c *Cmd) Run() error {
+	if err := strace.Trace(c.Cmd, func(t strace.Task, r *strace.TraceRecord) error {
+		ret := c.handleEvent(t, r, c.events)
+		return ret
+	}); err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	return nil
+}
+
+// AddActions creates an []event as defined by a possibly empty set of actions, and
+// installs it in the Cmd.
+// The action is a string, with comma-separated fields,
+// designed to be convenient to be used on a command line.
+//
+// Actions have 3 fields.
+//
+// The first is a regexp to match against the system call name, e.g.,
+// E.*imeofday. The first character, E or X,
+// indicates Entry or eXit.
+//
+// The second parameter indications an action to take. Currently there
+// are only two: error, and log.
+//
+// The third value indicates a value to return. It can be empty.
+// For the error action, the third parameter
+// indicates the error to return. For the log action, the third
+// parameter is currently unused.
+
+func (c *Cmd) AddActions(actions ...string) error {
+	var events []*event
+	for i, a := range actions {
+		f := strings.Split(a, ",")
+		if len(f) != 3 {
+			return fmt.Errorf("%d of actions: %q needs to have 3 fields, has %d(%v)", i, a, len(f), f)
+		}
+		if _, ok := allActions[f[1]]; !ok {
+			return fmt.Errorf("%q of actions %v: unknown action, not one of %q", f[0], f, allActions)
+		}
+		if check := findEvent(events, f[0]); check != nil {
+			return fmt.Errorf("%q of actions %v: repeat action, already %q", f[0], f, check)
+		}
+		var value int
+		if len(f[2]) > 0 {
+			v, err := strconv.Atoi(f[2])
```

Comment:
maybe this is me not understanding, but I usually prefer to use parseint with a base of 0 since that handles decimal and hex. Maybe no one actually thinks of this return value in hex though, so don't mind me in that case then.

Reply:
ah, oops. Yeah, fixed.